software/controller/lib/hal/gpio.h

```diff
@@ -80,30 +85,83 @@ enum class AlternativeFunction : uint32_t {
 
 void enable_all_clocks();
 
-void pin_mode(Port port, uint8_t pin, PinMode mode);
-
-void output_type(Port port, uint8_t pin, OutType output_type);
-
-void output_speed(Port port, uint8_t pin, OutSpeed speed);
-
-// Many GPIO pins can be repurposed with an alternate function
-// See Table 17 and 18 [DS] for alternate functions
-// See [RM] 8.4.9 and 8.4.10 for GPIO alternate function selection
-void alternate_function(Port port, uint8_t pin, AlternativeFunction func);
+// Abstract pin. Note that for some types of pin (most notably AlternateFunction), we only need to
+// instantiate the pin so the registers are properly set, but we don't really need to keep a
+// reference to the pin itself
+// Also note that because the instantiation of a pin writes to registers, this needs to happen
+// after the Hal has been properly initialized.
+class Pin {
+ public:
+  Pin(Port port, uint8_t pin, PinMode mode);
+
+ protected:
+  Port port_;
+  uint8_t pin_;
+  // helper functions to set properties of the pin (writes them to registers).
+  void output_type(OutType output_type);              // for output or alternate pins
+  void output_speed(OutSpeed speed);                  // for output or alternate pins
+  void pull_type(PullType pull);                      // for input or alternate pins
+  void alternate_function(AlternativeFunction func);  // for analog output or alternate pins
+};
 
-// Set a specific output pin
-void set_pin(Port port, uint8_t pin);
+// Output pin, which can be set to 1 or cleared to 0
+class DigitalOutputPin : public Pin {
+ public:
+  DigitalOutputPin(Port port, uint8_t pin, bool start_high = false, OutSpeed speed = OutSpeed::Slow,
+                   OutType type = OutType::PushPull);
+  void set();
+  void clear();
+#if !defined(BARE_STM32)
```

Comment:
Still somehow not great to have these `ifdefs` in the abstract header. Ideally, I feel like these Pin classes could be further subclassed as some mock classes specifically for unit tests, but I suspect they end up somehow deep within other classes being tested? We can defer solving this to another PR, but just want to take note of it already. We could throw a `TODO` in here.

Reply:
Agreed, though they are nested within actuators and sensors, I will make an effort to disentangle those in a next PR.
left a TODO.